### ISSUE TITLE
Tests fail behind proxy

### ISSUE THREAD

**Issue by petdance** [petdance/test-www-mechanize#34]
From https://code.google.com/p/www-mechanize/issues/detail?id=222

Reported by alex.hartmaier, Oct 27, 2011

What steps will reproduce the problem?
1. try to install Test::WWW::Mechnanize when behind a proxy without direct internet access having set the http_proxy env var set correctly.

What is the expected output? What do you see instead?

```
Installing Test::WWW::Mechanize (1.36)
Running [/home/nac/perl5/perlbrew/perls/perl-5.14.1/bin/perl /home/nac/perl5/perlbrew/perls/perl-5.14.1/bin/cpanp-run-perl /home/nac/.cpanplus/5.14.1/build/Test-WWW-Mechanize-1.36/Makefile.PL]...
Checking if your kit is complete...
Looks good
Writing Makefile for Test::WWW::Mechanize
Writing MYMETA.yml and MYMETA.json
Running [/usr/bin/make test]...
PERL_DL_NONLAZY=1 /home/nac/perl5/perlbrew/perls/perl-5.14.1/bin/perl "-MExtUtils::Command::MM" "-e" "test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
# Testing Test::WWW::Mechanize 1.36, with WWW::Mechanize 1.70, LWP 6.03, Perl 5.014001, /home/nac/perl5/perlbrew/perls/perl-5.14.1/bin/perl
t/00load.t ................. ok
Can't locate HTML/Lint.pm in @INC (@INC contains: /home/nac/.cpanplus/5.14.1/build/Test-WWW-Mechanize-1.36/blib/lib /home/nac/.cpanplus/5.14.1/build/Test-WWW-Mechanize-1.36/blib/arch /home/nac/perl5/perlbrew/perls/perl-5.14.1/lib/site_perl/5.14.1/x86_64-linux-thread-multi /home/nac/perl5/perlbrew/perls/perl-5.14.1/lib/site_perl/5.14.1 /home/nac/perl5/perlbrew/perls/perl-5.14.1/lib/5.14.1/x86_64-linux-thread-multi /home/nac/perl5/perlbrew/perls/perl-5.14.1/lib/5.14.1) at t/autolint.t line 8.
BEGIN failed--compilation aborted at t/autolint.t line 8.
t/autolint.t ............... 
Dubious, test returned 2 (wstat 512, 0x200)
No subtests run 
t/click_ok.t ............... ok
t/content_contains.t ....... ok
t/content_lacks.t .......... ok
t/follow_link_ok.t ......... ok
t/followable_links.t ....... ok
t/get_ok-parms.t ........... ok

#   Failed test 'Fails to get nonexistent URI and reports failure'
#   at t/get_ok.t line 51.
# STDERR is:
# #   Failed test 'Try to get bad URL'
# #   at t/get_ok.t line 50.
# # 502
# # Proxy Error ( The host was not found. )
# 
# not:
# /#\s+Failed\ test.*?\n?.*?at\ t\/get_ok\.t line 50.*\n?/
# 
# # 500
# 
# # Can't connect to wango.nonexistent.xx-only-testing:80 (Bad hostname)
# 
# as expected
t/get_ok.t ................. 
Failed 1/11 subtests 
t/has_tag.t ................ ok
t/head_ok-parms.t .......... ok

#   Failed test 'Fails to HEAD nonexistent URI and reports failure'
#   at t/head_ok.t line 51.
# STDERR is:
# #   Failed test 'Try to HEAD bad URL'
# #   at t/head_ok.t line 50.
# # 502
# # Proxy Error ( The host was not found. )
# 
# not:
# /#\s+Failed\ test.*?\n?.*?at\ t\/head_ok\.t line 50.*\n?/
# 
# # 500
# 
# # Can't connect to blahblablah.xx-nonexistent.foo:80 (Bad hostname)
# 
# as expected
t/head_ok.t ................ 
Failed 1/11 subtests 
t/html_lint_ok.t ........... skipped: HTML::Lint is not installed, cannot test html_lint_ok
t/lacks_uncapped_inputs.t .. ok
t/link_content.t ........... ok
t/link_status.t ............ ok
t/links_ok.t ............... ok
t/new.t .................... ok
t/page_links_content.t ..... ok
t/page_links_ok.t .......... ok
t/pod-coverage.t ........... skipped: Test::Pod::Coverage 0.08 required for testing POD coverage
t/pod.t .................... ok
t/put_ok.t ................. ok
t/stuff_inputs.t ........... ok
t/submit_form_ok.t ......... ok
t/text_contains.t .......... ok

Test Summary Report
-------------------
t/autolint.t             (Wstat: 512 Tests: 0 Failed: 0)
  Non-zero exit status: 2
  Parse errors: No plan found in TAP output
t/follow_link_ok.t       (Wstat: 0 Tests: 7 Failed: 0)
  TODO passed:   5-6
t/get_ok.t               (Wstat: 0 Tests: 11 Failed: 1)
  Failed test:  9
t/head_ok.t              (Wstat: 0 Tests: 11 Failed: 1)
  Failed test:  9
Files=26, Tests=231, 11 wallclock secs ( 0.12 usr  0.22 sys +  5.58 cusr  1.65 csys =  7.57 CPU)
Result: FAIL
Failed 3/26 test programs. 2/231 subtests failed.
make: *** [test_dynamic] Error 255
```

What version of the product are you using? On what operating system?

version 1.36

Please provide any additional information below.
## Oct 25, 2012

mmlange

I would also add that for testing-only of Test::WWW::Mechanize, using the http_proxy (or ftp_proxy) variables for access to a server spawned on 127.0.0.1 is probably not a good idea.

My output is:

```
#   Failed test 'GET http://127.0.0.1:34903/form.html'
#   at t/click_ok.t line 21.
#503
# Service Unavailable
```

...which is expected since the test asked my proxy to fetch content from itself.
## Apr 4, 2014

three18ti

I experienced this issue as well.  I was able to work around it by setting my $http_proxy variable to ''.

```
$ cpanm -v Test::WWW::Mechanize
    #   Failed test 'GET http://127.0.0.1:26366/form.html'
    #   at t/click_ok.t line 21.
    #504
    # Gateway Timeout
^C
$ cd  ~/.cpanm/latest-build/Test-WWW-Mechanize-1.44
$ make
$ make test
# test pass
$ make install
```

Then I was able to install other modules that require Test::WWW::Mechanize

**Comment by glasswalk3r** [comment]
Confirmed in my smoker too.

I setup a proxy to force the smoker failing trying to reach for the internet when any distribution is missing on my local Mini::CPAN.

If I sent an empty value for `http_proxy` on the CPAN client, it works as expected (report to be sent in the next few minutes).

I even tried a distroprefs, but CPAN just ignores changing `http_proxy`:

```yaml
---
comments: trying to force http_proxy to be changed temporarily
match:
  distribution: ^PETDANCE/Test-WWW-Mechanize
cpanconfig:
  http_proxy: ''
```

**Comment by sudo-batman** [comment]
I am also facing the similar issue behind proxy-
version- 1.52 on Ubuntu 16.04

```
Running make test
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/00-load.t ................ 1/1 # Testing Test::WWW::Mechanize 1.52, with WWW::Mechanize 1.75, LWP 6.15, Test::More 1.302141, Perl 5.022001, /usr/bin/perl
# Optional HTML::Lint not found. Install it to use additional features.
# Optional HTML::Tidy5 not found. Install it to use additional features.
t/00-load.t ................ ok   
t/autolint.t ............... skipped: HTML::Lint 2.20 is not installed, cannot test autolint
t/autotidy.t ............... skipped: Optional HTML::Tidy5 1.00 is not installed, cannot test autotidy.
t/button_exists.t .......... ok   
t/click_ok.t ............... # Test server http://127.0.0.1:20064 as PID 11725
t/click_ok.t ............... 2/6 
#   Failed test 'GET http://127.0.0.1:20064/form.html'
#   at t/click_ok.t line 21.
# 503
# Connect failed
Can't call method "click" on an undefined value at /usr/share/perl5/WWW/Mechanize.pm line 1742.
# Looks like your test exited with 255 just after 2.
```

And it stuck here unless I stop it manually.

**Comment by petdance** [comment]
I believe this is fixed by #55
